File: wifi-bot.cpp
#include <wiringPi.h>
#include <iostream>
#include <thread>
#include <string>
#include "wifi-bot.h"
#include <SFML/Window.hpp>

int moveForward()
{
	digitalWrite(FRONT_RIGHT_FORWARD, HIGH);
	digitalWrite(FRONT_LEFT_FORWARD, HIGH);
	digitalWrite(BACK_RIGHT_FORWARD, HIGH);
	digitalWrite(BACK_LEFT_FORWARD, HIGH);
	
	return FORWARD_MOVE;
}

int moveBackward()
{
	digitalWrite(FRONT_RIGHT_BACKWARD, HIGH);
	digitalWrite(FRONT_LEFT_BACKWARD, HIGH);
	digitalWrite(BACK_RIGHT_BACKWARD, HIGH);
	digitalWrite(BACK_LEFT_BACKWARD, HIGH);
	
	return BACKWARD_MOVE;
}

int rotateRight()
{
	digitalWrite(FRONT_LEFT_FORWARD, HIGH);
	digitalWrite(BACK_LEFT_FORWARD, HIGH);
	digitalWrite(FRONT_RIGHT_BACKWARD, HIGH);
	digitalWrite(BACK_RIGHT_BACKWARD, HIGH);
	
	return RIGHT_ROTATE;
}

int rotateLeft()
{
	digitalWrite(FRONT_RIGHT_FORWARD, HIGH);
	digitalWrite(BACK_RIGHT_FORWARD, HIGH);
	digitalWrite(FRONT_LEFT_BACKWARD, HIGH);
	digitalWrite(BACK_LEFT_BACKWARD, HIGH);
	
	return LEFT_ROTATE;
}

void stopMove(int move_code)
{
	if(move_code == FORWARD_MOVE)
	{
		digitalWrite(FRONT_RIGHT_FORWARD, LOW);
		digitalWrite(FRONT_LEFT_FORWARD, LOW);
		digitalWrite(BACK_RIGHT_FORWARD, LOW);
		digitalWrite(BACK_LEFT_FORWARD, LOW);
	}
	else if(move_code == BACKWARD_MOVE)
	{
		digitalWrite(FRONT_RIGHT_BACKWARD, LOW);
		digitalWrite(FRONT_LEFT_BACKWARD, LOW);
		digitalWrite(BACK_RIGHT_BACKWARD, LOW);
		digitalWrite(BACK_LEFT_BACKWARD, LOW);
	}
	else if(move_code == RIGHT_ROTATE)
	{
		digitalWrite(FRONT_LEFT_FORWARD, LOW);
		digitalWrite(BACK_LEFT_FORWARD, LOW);
		digitalWrite(FRONT_RIGHT_BACKWARD, LOW);
		digitalWrite(BACK_RIGHT_BACKWARD, LOW);
	}
	else if(move_code == LEFT_ROTATE)
	{
		digitalWrite(FRONT_RIGHT_FORWARD, LOW);
		digitalWrite(BACK_RIGHT_FORWARD, LOW);
		digitalWrite(FRONT_LEFT_BACKWARD, LOW);
		digitalWrite(BACK_LEFT_BACKWARD, LOW);
	}
}

int main()
{
	
	using std::cout;
	using std::endl;
	int last_input = -1; //null
	bool exit;
	
	wiringPiSetup();
	
	pinMode(FRONT_RIGHT_FORWARD, OUTPUT);
	pinMode(FRONT_RIGHT_BACKWARD, OUTPUT);
	pinMode(FRONT_LEFT_FORWARD, OUTPUT);
	pinMode(FRONT_LEFT_BACKWARD, OUTPUT);
	pinMode(BACK_RIGHT_BACKWARD, OUTPUT);
	pinMode(BACK_RIGHT_FORWARD, OUTPUT);
	pinMode(BACK_LEFT_FORWARD, OUTPUT);
	pinMode(BACK_LEFT_BACKWARD, OUTPUT);
	
	
	while(!exit)
	{
		
		if(sf::Keyboard::isKeyPressed(sf::Keyboard::Up))
		{
			last_input = moveForward();
		}
		else if(sf::Keyboard::isKeyPressed(sf::Keyboard::Down))
		{
			last_input = moveBackward();
		}
		else if(sf::Keyboard::isKeyPressed(sf::Keyboard::Right))
		{
			last_input = rotateRight();
		}
		else if(sf::Keyboard::isKeyPressed(sf::Keyboard::Left))
		{
			last_input = rotateLeft();
		}
		else if(sf::Keyboard::isKeyPressed(sf::Keyboard::Escape))
		{
			exit = true;
		}
		else
		{
			stopMove(last_input);
		}
		
		delay(10);
		
	}
	
	cout << "Exiting program" << endl;
	
	return 0;
	
}

File: wifi-bot.py
import curses
import RPi.GPIO as GPIO

GPIO.setmode(GPIO.BCM)

FRONT_RIGHT_BACKWARD = 2
FRONT_RIGHT_FORWARD = 3
FRONT_LEFT_BACKWARD = 13
FRONT_LEFT_FORWARD = 17
BACK_RIGHT_BACKWARD = 27
BACK_RIGHT_FORWARD = 22
BACK_LEFT_BACKWARD = 10
BACK_LEFT_FORWARD = 9

GPIO.setup(FRONT_RIGHT_BACKWARD, GPIO.OUT)
GPIO.setup(FRONT_RIGHT_FORWARD, GPIO.OUT)
GPIO.setup(FRONT_LEFT_BACKWARD, GPIO.OUT)
GPIO.setup(FRONT_LEFT_FORWARD, GPIO.OUT)
GPIO.setup(BACK_RIGHT_BACKWARD, GPIO.OUT)
GPIO.setup(BACK_RIGHT_FORWARD, GPIO.OUT)
GPIO.setup(BACK_LEFT_BACKWARD, GPIO.OUT)
GPIO.setup(BACK_LEFT_FORWARD, GPIO.OUT)

def stopMove():
    GPIO.output(2,False)
    GPIO.output(3,False)
    GPIO.output(13,False)
    GPIO.output(17,False)
    GPIO.output(27,False)
    GPIO.output(22,False)
    GPIO.output(10,False)
    GPIO.output(9,False)

def moveForward():
    GPIO.output(FRONT_RIGHT_FORWARD, True)
    GPIO.output(FRONT_LEFT_FORWARD, True)
    GPIO.output(BACK_RIGHT_FORWARD, True)
    GPIO.output(BACK_LEFT_FORWARD, True)

def moveBackward():
    GPIO.output(FRONT_RIGHT_BACKWARD, True)
    GPIO.output(FRONT_LEFT_BACKWARD, True)
    GPIO.output(BACK_RIGHT_BACKWARD, True)
    GPIO.output(BACK_LEFT_BACKWARD, True)

def rotateRight():
    GPIO.output(FRONT_LEFT_FORWARD, True)
	GPIO.output(BACK_LEFT_FORWARD, True)
	GPIO.output(FRONT_RIGHT_BACKWARD, True)
	GPIO.output(BACK_RIGHT_BACKWARD, True)

def rotateLeft():
    GPIO.output(FRONT_RIGHT_FORWARD, True)
	GPIO.output(BACK_RIGHT_FORWARD, True)
	GPIO.output(FRONT_LEFT_BACKWARD, True)
	GPIO.output(BACK_LEFT_BACKWARD, True)



screen = curses.initscr()
curses.noecho()
curses.cbreak()
screen.keypad(True)

try:
    while True:
        char = screen.getch()
        
        if char == ord('q'):
            break
        elif char == curses.KEY_UP:
            moveForward()
            break
        elif char == curses.KEY_DOWN:
            moveBackward()
            break
        elif char == curses.KEY_RIGHT:
            rotateRight()
            break
        elif char == curses.KEY_LEFT:
            rotateLeft()
            break
        elif char == 10:
            stopMove()
            break
finally:
    curses.nocbreak(); screen.keypad(0); curses.echo()
    curses.endwin()
    GPIO.cleanup()


